docs/topics/using-builders-with-builder-inference.md

```diff
@@ -0,0 +1,253 @@
+[//]: # (title: Using builders with builder type inference)
+
+Kotlin supports _builder type inference_ (or builder inference), which is a type inference flavour that comes in handy 
+when writing generic builders. It helps the compiler infer the type arguments of a builder call based on the type information 
+about other calls inside its lambda argument.
+
+Consider the example of the [`buildMap()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-map.html) 
+usage:
+
+```kotlin
+fun addEntryToMap(baseMap: Map<String, Number>, additionalEntry: Pair<String, Int>?) {
+    val myMap = buildMap {
+        putAll(baseMap)
+        if (additionalEntry != null) {
+            put(additionalEntry.first, additionalEntry.second)
+        }
+    }
+}
+```
+
+Here we don't have enough type information to infer type arguments in a regular way but builder inference can
+analyze the calls inside the lambda argument. Based on the type information about `putAll()` and `put()` calls, 
+the compiler can automatically infer type arguments of the `buildMap()` call into `String` and `Number`.
+
+As you can see, builder inference allows omitting type arguments while using generic builders.
+
+## Writing own builders
+
+### Requirements for enabling builder inference
+
+> Before Kotlin 1.6.0, enabling builder inference for a builder function required the [`@BuilderInference`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-builder-inference/) 
+> annotation to be present on a builder lambda parameter. In 1.6.0, you can omit the annotation on condition of using
+> the compiler option `-Xenable-builder-inference` both on your and client's side.
+> 
+{type="note"}
+
+To let builder inference work for your own builder, make sure its declaration has a builder lambda parameter of a 
+function type with receiver. There are also two requirements to a receiver type:
+
+1. It should use the type arguments that builder inference is supposed to infer. For example:
+   ```kotlin
+   fun <V> buildList(builder: MutableList<V>.() -> Unit) { ... }
+   ```
+   
+   > Note that passing type parameter's type directly like `fun <T> myBuilder(builder: T.() -> Unit)` is not yet supported.
+   > 
+   {type="note"}
+
+2. It should provide public members that contain the corresponding type parameters in their signature. For example:
```

Comment:
Sorry, it seems I haven't mentioned it in the internal doc, but extensions are also possible to use:
```
class ItemHolder<T> {}

fun <T> ItemHolder<T>.foo(x: T) {}

fun <T> itemHolderBuilder(builder: ItemHolder<T>.() -> Unit): ItemHolder<T> =
    ItemHolder<T>().apply(builder)

fun test(s: String) {
    val itemHolder1 = itemHolderBuilder { // Type of itemHolder1 is ItemHolder<String>
        foo(s)
    }
}
```